Fix Cash App Pay token reuse issues (with subscriptions)

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes #3171

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

This PR enables the use of Cash App Pay when paying for subscriptions with free trials and as an option when changing a subscription's payment method.

For that, I have updated the usage of the Cash App Pay confirm method. The wallet hash generation method includes a new parameter (the payment intent type).

**Known issue**: The subscription extension adds a "success message" before the redirect is completed, which can be confusing for customers.

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

- Checkout to this branch on your local environment (`fix/cashapp-token-reuse2`)
- Run `npm install`, `npm run build:webpack` and `npm run up`

(the following testing guide was extracted from the [original issue](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3171))

**Free trial** 
1. Change your Stripe Plugin credentials to a US-based account.
3. Change your store currency to USD.
4. Enable Cash App Pay from the Stripe payment methods screen.
5. Enable WooCommerce Subscriptions. 
6. Create a Subscription product with a free trial. 
7. Add the product to your cart. 
8. Go to the checkout
9. Confirm that you can select Cash App Pay as a payment method
10. Purchase the subscription using Cash App - this will create a token. 
11. Add the free trial product to your cart again. 
12. Confirm that you can complete the sign-up using the saved payment method.

**Changing payment method** 
1. Purchase a subscription product using a card.
1. Go to the **My Account → Subscriptions**
2. View your active subscription. 
3. Click the "Change payment" button.
6. Confirm that the Cash App Pay method is now available.
7. Confirm that you can change the subscription's payment to **a saved cash app token**. 

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Fix Cash App Pay reuse of tokens

wjrosa committed:
Fix cash app confirm intent call

wjrosa committed:
Fix parsing of wallet URL partials

wjrosa committed:
Removing additional log

wjrosa committed:
Enable selecting Cash App when changing subscription payment methods

wjrosa committed:
Possible redirect fix commented out

wjrosa committed:
Merge branch 'develop' into fix/cashapp-token-reuse2

wjrosa committed:
Changelog and readme updates

wjrosa committed:
Merge branch 'develop' into fix/cashapp-token-reuse2

wjrosa committed:
Uncommenting out wallet confirmation function updates

wjrosa committed:
Fix tests

wjrosa committed:
Updating docs related to the wallet redirect hash

james-allan commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -834,6 +834,14 @@ private function process_payment_with_deferred_intent( int $order_id ) {
+
+				return [
+					'result'   => 'success',
+					'redirect' => '',
```
```suggestion
					'redirect' => $this->get_return_url( $order ),
```

Without a redirect, using a saved cash app token for an order that doesn't need payment, just leaves the customer on the checkout page.

james-allan reviewed (changes requested):
Hey @wjrosa. I've been following some general Subscription testing instructions. I've listed them all below. I ran into a couple of issues, I'll follow up and finish testing shortcode checkouts tomorrow or later this week. 

**BLOCK CHECKOUT**
- **Purchase subscription**
   - [x] New token
   - [x] Existing token
- **Free trial subscription**
   - [x] New token
   - [ ] ❗ Existing token
       - Selecting an existing token and attempting to sign up to a free trial ~~just leaves the customer on the checkout page~~ leaves the subscription without a stored payment method. There's more to this as well. I'll leave a comment below. *
- **Mixed checkout (simple + subscriptions)**
   - [x] New token
   - [x] Existing token
- **Multiple Subscriptions**
   - [x] New token
   - [x] Existing token

**SHORTCODE CHECKOUT**
- **Purchase subscription**
   - [x] New token
   - [x] Existing token
- **Free trial subscription**
   - [ ] New token
   - [ ] Existing token
- **Mixed checkout (simple + subscriptions)**
   - [x] New token
   - [x] Existing token
- **Multiple Subscriptions**
   - [x] New token
   - [x] Existing token

**OTHER FEATURES**
- **Renew subscriptions**
   - [x] New token
   - [x] Existing token
- **Change subscription payment**
   - ~~❗ I cannot get Cash App to show as an option on the change payment method screen. ([screenshot](https://github.com/user-attachments/assets/07d9075c-00a9-43b3-8f19-22a4d7d172b6))~~
   - [x] New token
   - [x] Existing token
- [x] Add payment method via My Account.
   - ~~❗ The Cash App modal isn't shown. There's a browser console error and [this error](https://github.com/user-attachments/assets/4bb26dac-b1e0-45be-b3f5-76d262874c00) in Stripe.~~ This is not related to this PR. Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3311
- **Failed/manual renewal - Block Checkout**
   - [x] New token
   - [x] Existing token
- **Failed/manual renewal - Shortcode Checkout**
   - [x] New token
   - [x] Existing token

\* **Free trial - existing token issue.** Even if I fixed the redirect problem, there's another issue where the subscription's payment method isn't set to the Cash App token. This is because the return happens so early in the `process_payment_with_deferred_intent()` function that things like saving the payment method information doesn't happen.

wjrosa committed:
Update includes/payment-methods/class-wc-stripe-upe-payment-gateway.php

Co-authored-by: James Allan <james.allan@automattic.com>

wjrosa committed:
Merge branch 'develop' into fix/cashapp-token-reuse2

wjrosa committed:
Fix Cash App Pay token inclusion

wjrosa commented on includes/class-wc-stripe-intent-controller.php:
```diff
@@ -956,7 +956,7 @@ public function is_mandate_data_required( $selected_payment_type, $is_using_save
 		$request = [
 			'payment_method'       => $payment_information['payment_method'],
-			'payment_method_types' => $payment_information['payment_method_types'],
+			'payment_method_types' => $payment_information['payment_method_types'] ?? [ $payment_information['selected_payment_type'] ],
```
Hi @james-allan ! Your change [here](https://github.com/woocommerce/woocommerce-gateway-stripe/commit/3929b33faa6aedf7864b531e4ce845ebdaa9f186) broke this part when adding Cash App Pay tokens. `$payment_information['payment_method_types']` does not exist in this context. I have added the previous version as a fallback to fix it.

james-allan replied:
👍 Thanks for the heads up. I also changed it back to fallback to the `selected_payment_type` like you did. https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3311/commits/ebcd3b3604638c20673f6d59678159a0f780b86a.

wjrosa commented:
@james-allan 
> * Free trial - existing token issue. Even if I fixed the redirect problem, there's another issue where the subscription's payment method isn't set to the Cash App token. This is because the return happens so early in the process_payment_with_deferred_intent() function that things like saving the payment method information doesn't happen.

I believe the latest changes fixed this issue.

> Change subscription payment
❗ I cannot get Cash App to show as an option on the change payment method screen. ([screenshot](https://github.com/user-attachments/assets/07d9075c-00a9-43b3-8f19-22a4d7d172b6))

I had the same issue and did some debugging. The issue happens on `togglePaymentMethodForCountry`. For some reason `getStripeServerData()?.customerData?.billing_country` is not being populated. My country is set to US (on the checkout and on the Stripe side as a customer), but `customerData` is undefined.

It looks like this property is set by calling `useCustomerData`. Not sure if we need to call it manually on the change payment method page 🤔  (it is not called there). Still, does not seem directly related to this PR.

wjrosa committed:
Turning customer country available when changing a subscription payment method so Cash App can be selected

wjrosa commented:
I actually fixed the above on https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3263/commits/64b20d5edf7179fae8fda283b3b0f3e0fe865dd8 @james-allan 👍

wjrosa committed:
Merge branch 'develop' into fix/cashapp-token-reuse2

wjrosa committed:
Fix redirect when using a new token

wjrosa committed:
Removing unnecessary check when saving method to subscription

wjrosa commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -1647,10 +1661,7 @@ public function set_payment_method_title_for_order( $order, $payment_method_type
 		$order->save();
 
 		// Update the subscription's purchased in this order with the payment method ID.
-		if ( isset( $this->payment_methods[ $payment_method_type ] ) ) {
```
This check is not necessary since we have the same thing above.

wjrosa commented:
@james-allan I thought this was not working:

> Change subscription payment
> * New token

But it is. It seems that we have an issue with this cache:
```
self::PAYMENT_METHODS_TRANSIENT_KEY . $payment_method_type . $this->get_id()
```
It is not saving the new Cash App method there. Consequently, when displaying the method, it is shown as "N/A".

But it's not related to this change, I guess.

wjrosa commented:
hi @james-allan ! I tested the other flows from your checklist myself, and I guess it is all good. One thing that I noticed is that when using an existing token, the subscription is set to "active", while with a new token, it is set to "pending". Is this expected? 🤔 I thought both would require a webhook call to process

wjrosa committed:
Merge branch 'develop' into fix/cashapp-token-reuse2

wjrosa committed:
Revert unnecessary change

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Re-including tests removed by mistake

james-allan commented:
> > Free trial - existing token issue. Even if I fixed the redirect problem, there's another issue where the subscription's payment method isn't set to the Cash App token. This is because the return happens so early in the process_payment_with_deferred_intent() function that things like saving the payment method information doesn't happen.

> I believe the latest changes fixed this issue.

@wjrosa this flow is causing an issue for me. If you add a subscription with a free trial to your cart and then use a saved Cash App token, the resulting subscription will have missing payment method token information.


<p align="center">
<img width="1064" alt="Screenshot 2024-08-13 at 2 53 17 PM" src="https://github.com/user-attachments/assets/7ff5f8cf-9a0c-4a25-b7dd-e95392c47229">
</br>
<sup>Empty payment method information on <strong>My Account → Subscriptions</strong> list.</sup>
</p> 

<p align="center">
<img width="500" alt="Screenshot 2024-08-13 at 2 55 20 PM" src="https://github.com/user-attachments/assets/bcd34d58-ff8a-4816-982b-10e48263be06">
</br>
<sup>Empty payment method information on <strong>WooCommerce → Subscriptions (edit)</strong> screen</sup>
</p> 

To fix this we'd need to update [this section](https://github.com/woocommerce/woocommerce-gateway-stripe/blame/fix/cashapp-token-reuse2/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L849-L856) of the code. If you purchase a free subscription with a saved token, the payment_complete and early return occurs without storing the token to the order. We essentially need to make sure we still save the token to the order too. re: 

```php
$this->maybe_update_source_on_subscription_order(
	$order,
	(object) [
		'payment_method' => $payment_information['payment_method'],
		'customer'       => $payment_information['customer'],
	],
	$this->get_upe_gateway_id_for_order( $upe_payment_method )
);
```

james-allan commented:
> One thing that I noticed is that when using an existing token, the subscription is set to "active", while with a new token, it is set to "pending". Is this expected? 🤔 I thought both would require a webhook call to process

Yeah that seems a little off. Are you sure one of them is just going straight to active? My webhooks are so quickly processed that I don't notice any difference. But yes, if you're using a saved payment method or a new token to pay for a new subscription that has an upfront cost I'd expect both of them to involve a webhook process.

In my case when I use a new or existing token by the time I land on the order received page the subscription is active.

Now that I think on it, perhaps what you're seeing is the difference between a setup intent and just payment intent. 🤷 In my testing though both result in a `payment_intent.succeeded` and a `charge.succeeded` webhook.

james-allan reviewed (changes requested):
Changes look pretty good @wjrosa. I just left a comment about the subscriptions with a free trial being purchased with a saved payment method.

wjrosa committed:
Merge branch 'develop' into fix/cashapp-token-reuse2

wjrosa committed:
Fix missing token information when subscribing using a saved token

wjrosa committed:
Adding more doc blocks

wjrosa commented:
> > > Free trial - existing token issue. Even if I fixed the redirect problem, there's another issue where the subscription's payment method isn't set to the Cash App token. This is because the return happens so early in the process_payment_with_deferred_intent() function that things like saving the payment method information doesn't happen.
> 
> > I believe the latest changes fixed this issue.
> 
> @wjrosa this flow is causing an issue for me. If you add a subscription with a free trial to your cart and then use a saved Cash App token, the resulting subscription will have missing payment method token information.
> 
> <img alt="Screenshot 2024-08-13 at 2 53 17 PM" width="1064" src="https://private-user-images.githubusercontent.com/8490476/357279514-7ff5f8cf-9a0c-4a25-b7dd-e95392c47229.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1NTczOTAsIm5iZiI6MTcyMzU1NzA5MCwicGF0aCI6Ii84NDkwNDc2LzM1NzI3OTUxNC03ZmY1ZjhjZi05YTBjLTRhMjUtYjdkZC1lOTUzOTJjNDcyMjkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTNUMTM1MTMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2RiMDBhYmU0ZjZlYjQ1NDZlMzg2ZGQwYTM5MTkwOWYxNzhlMmIxNWRmNTNmMmM0NjIyMmQwMWE5Y2I4MTViZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.C5Dmo_uF9rhNlKXC4fLW6nnU4X8XloUY-liIDO4EzVc"> Empty payment method information on **My Account → Subscriptions** list.
> 
> <img alt="Screenshot 2024-08-13 at 2 55 20 PM" width="500" src="https://private-user-images.githubusercontent.com/8490476/357280020-bcd34d58-ff8a-4816-982b-10e48263be06.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1NTczOTAsIm5iZiI6MTcyMzU1NzA5MCwicGF0aCI6Ii84NDkwNDc2LzM1NzI4MDAyMC1iY2QzNGQ1OC1mZjhhLTQ4MTYtOTgyYi0xMGU0ODI2M2JlMDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTNUMTM1MTMwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmE5YWZhY2VjOTc3M2ZmMmU5OGFjYTU0YTU4MzhlOTI0MmUyNTI1ZGQyMDA3OTRmMDkyZTc3OGFiNDY2MWVmMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.hV_np5pdwS3N5s9FR7aaDSeYrzlEU6UaX_iP55Bua38"> Empty payment method information on **WooCommerce → Subscriptions (edit)** screen
> 
> To fix this we'd need to update [this section](https://github.com/woocommerce/woocommerce-gateway-stripe/blame/fix/cashapp-token-reuse2/includes/payment-methods/class-wc-stripe-upe-payment-gateway.php#L849-L856) of the code. If you purchase a free subscription with a saved token, the payment_complete and early return occurs without storing the token to the order. We essentially need to make sure we still save the token to the order too. re:
> 
> ```
> $this->maybe_update_source_on_subscription_order(
> 	$order,
> 	(object) [
> 		'payment_method' => $payment_information['payment_method'],
> 		'customer'       => $payment_information['customer'],
> 	],
> 	$this->get_upe_gateway_id_for_order( $upe_payment_method )
> );
> ```

Thanks, @james-allan ! I fixed this on https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3263/commits/b8bd1afac2d8d885901ab3b0cdf87baeaa828211

wjrosa commented:
> > One thing that I noticed is that when using an existing token, the subscription is set to "active", while with a new token, it is set to "pending". Is this expected? 🤔 I thought both would require a webhook call to process
> 
> Yeah that seems a little off. Are you sure one of them is just going straight to active? My webhooks are so quickly processed that I don't notice any difference. But yes, if you're using a saved payment method or a new token to pay for a new subscription that has an upfront cost I'd expect both of them to involve a webhook process.
> 
> In my case when I use a new or existing token by the time I land on the order received page the subscription is active.
> 
> Now that I think on it, perhaps what you're seeing is the difference between a setup intent and just payment intent. 🤷 In my testing though both result in a `payment_intent.succeeded` and a `charge.succeeded` webhook.

Any ideas where this status change to `active` is handled? Maybe that's something to do with the early return after calling `payment_complete()`.

wjrosa commented:
Still not sure where the issue is, but I did some debugging and saw that:
- When purchasing a product using an existing Cash App Pay token, we are calling `create_and_confirm_payment_intent`
- So, we are indeed processing the payment intent already
- Then getting inside this condition:
```
if ( $payment_needed ) {
				// Use the last charge within the intent to proceed.
				$charge = $this->get_latest_charge_from_intent( $payment_intent );

				// Only process the response if it contains a charge object. Intents with no charge require further action like 3DS and will be processed later.
				if ( $charge ) {
					$this->process_response( $charge, $order );
				}
```
- The `$charge` value will be from the intent we just processed before
- As it is `captured` and not `pending` but `succeeded`, it will get into this condition:
```
			if ( 'succeeded' === $response->status ) {
```
- Making the order `completed`, and triggering the hook to change the subscription status along it (to `active`)

So, what should we do in this case? (when using a saved Cash App Pay token for a paid subscription)
1. Leave it as is, as the intent is indeed being processed, and the webhook call will probably ignore this order/subscription
2. Skip this line and also return early:
```
// Create a payment intent, or update an existing one associated with the order.
$payment_intent = $this->process_payment_intent_for_order($order, $payment_information);
```
(but it may break something as we will have only the setup intent when the order/subscription is processed by the webhook call)
3. Return early before this check:
```
			if ( $payment_needed ) {
				// Use the last charge within the intent to proceed.
				$charge = $this->get_latest_charge_from_intent( $payment_intent );

				// Only process the response if it contains a charge object. Intents with no charge require further action like 3DS and will be processed later.
				if ( $charge ) {
					$this->process_response( $charge, $order );
				}
```
4. Force it to fall into one of the next conditions if it makes sense: `elseif ( $this->is_changing_payment_method_for_subscription() )` or `elseif ( in_array( $payment_intent->status, self::SUCCESSFUL_INTENT_STATUS, true ) )` (by looking I would say only the last would make a bit sense)

wjrosa committed:
Improving code readability

wjrosa committed:
Merge branch 'develop' into fix/cashapp-token-reuse2

wjrosa committed:
Fix lint issues

wjrosa committed:
Merge branch 'develop' into fix/cashapp-token-reuse2

wjrosa committed:
Changelog fix

wjrosa committed:
Minor doc updates

wjrosa commented:
@james-allan I did some additional analysis today for the status difference I mentioned earlier (after your fix for live payments), and I think the issue might be the `$order->payment_complete();`

### Non-saved token
- Order is kept as `pending`
- Confirmation is made on the frontend (QR code scan)

### Saved token
- Order is set to `completed` due to the call to `$order->payment_complete();`
- Nothing else is done

So:
- Should we have a QR code scan for saved tokens? From our previous conversations, I think it does not make sense (or did it change due to the live payments fix?)
- Should we just remove the call to `payment_complete` here instead?

wjrosa committed:
Minor variable name change

wjrosa commented:
I just tested this on a JN site, and the webhook call worked. So, maybe we don't need to do anything else here.

james-allan commented:
> Should we have a QR code scan for saved tokens? From our previous conversations, I think it does not make sense (or did it change due to the live payments fix?)

Nah I don't think we do need to display the QR code because the PM has already been set up and we have a saved token. 

> I just tested this on a JN site, and the webhook call worked. So, maybe we don't need to do anything else here.

Ok I'll rerun through some tests and confirm this works as expected.

james-allan committed:
Merge branch 'develop' into fix/cashapp-token-reuse2

james-allan committed:
Remove changelog entry that slipped into this branch

james-allan reviewed (approved):
**BLOCK CHECKOUT**
- **Purchase subscription**
   - [x] New token
   - [x] Existing token
- **Free trial subscription**
   - [x] New token
   - [x] Existing token
- **Mixed checkout (simple + subscriptions)**
   - [x] New token
   - [x] Existing token
- **Multiple Subscriptions**
   - [x] New token
   - [x] Existing token

**OTHER FEATURES**
- **Renew subscriptions**
   - [x] New token
   - [x] Existing token
- **Change subscription payment**
   - [x] New token
   - [x] Existing token
- [x] Add payment method via My Account.
- **Failed/manual renewal - Block Checkout**
   - [x] New token
   - [x] Existing token

I also ran a basic test using a new token and a saved token on the shortcode checkout just to be sure.
